tests/server.py

```diff
@@ -165,27 +180,35 @@ def headers(self) -> Headers:
             h.addRawHeader(*i)
         return h
 
-    def writeHeaders(self, version, code, reason, headers):
+    def writeHeaders(
+        self, version: bytes, code: bytes, reason: bytes, headers: Headers
+    ) -> None:
         self.result["version"] = version
         self.result["code"] = code
         self.result["reason"] = reason
         self.result["headers"] = headers
 
-    def write(self, content: bytes) -> None:
-        assert isinstance(content, bytes), "Should be bytes! " + repr(content)
+    def write(self, data: bytes) -> None:
+        assert isinstance(data, bytes), "Should be bytes! " + repr(data)
 
         if "body" not in self.result:
             self.result["body"] = b""
 
-        self.result["body"] += content
+        self.result["body"] += data
+
+    def writeSequence(self, data: Iterable[bytes]) -> None:
+        for x in data:
+            self.write(x)
+
+    def loseConnection(self) -> None:
+        self.unregisterProducer()
+        self.transport.loseConnection()
 
     # Type ignore: mypy doesn't like the fact that producer isn't an IProducer.
-    def registerProducer(  # type: ignore[override]
-        self,
-        producer: Union[IPullProducer, IPushProducer],
-        streaming: bool,
-    ) -> None:
-        self._producer = producer
+    def registerProducer(self, producer: IProducer, streaming: bool) -> None:
+        # Ensure that the producer implements one or more of IPushProducer and IPullProducer.
+        assert not set(providedBy(producer)).isdisjoint({IPushProducer, IPullProducer})
+        self._producer = cast(Union[IPushProducer, IPullProducer], producer)
```

Comment:
Tests are failing now, e.g.
```
===============================================================================
Error: 
Traceback (most recent call last):
  File "/home/runner/work/synapse/synapse/tests/rest/media/v1/test_media_storage.py", line 396, in test_thumbnail_scale
    unable_to_thumbnail=self.test_image.unable_to_thumbnail,
  File "/home/runner/work/synapse/synapse/tests/rest/media/v1/test_media_storage.py", line 554, in _test_thumbnail
    Image.open(BytesIO(channel.result["body"]))
builtins.KeyError: 'body'

tests.rest.media.v1.test_media_storage.MediaRepoTests_1.test_thumbnail_scale
tests.rest.media.v1.test_media_storage.MediaRepoTests_2.test_thumbnail_scale
-------------------------------------------------------------------------------
```

I stuck some debug in and got a traceback:


```
2023-02-17 15:13:45+0000 [-] 2023-02-17 15:13:45,005 - synapse.rest.media.v1._base - 270 - DEBUG - GET-0 - Responding to media request with responder <synapse.rest.media.v1.media_storage.FileResponder object at 0x7f15029c3890>
2023-02-17 15:13:45+0000 [-] 2023-02-17 15:13:45,005 - synapse.rest.media.v1._base - 279 - WARNING - GET-0 - Failed to write to consumer: <class 'AssertionError'> 
	Traceback (most recent call last):
	  File "/home/dmr/workspace/synapse/synapse/rest/media/v1/_base.py", line 274, in respond_with_responder
	    await responder.write_to_consumer(request)
	          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
	  File "/home/dmr/workspace/synapse/synapse/rest/media/v1/media_storage.py", line 330, in write_to_consumer
	    FileSender().beginFileTransfer(self.open_file, consumer)
	  File "/home/dmr/.cache/pypoetry/virtualenvs/matrix-synapse-YNqvS2_4-py3.11/lib/python3.11/site-packages/twisted/protocols/basic.py", line 886, in beginFileTransfer
	    self.consumer.registerProducer(self, False)
	  File "/home/dmr/.cache/pypoetry/virtualenvs/matrix-synapse-YNqvS2_4-py3.11/lib/python3.11/site-packages/twisted/web/http.py", line 1072, in registerProducer
	    self.channel.registerProducer(producer, streaming)
	  File "/home/dmr/workspace/synapse/tests/server.py", line 210, in registerProducer
	    assert not set(providedBy(producer)).isdisjoint({IPushProducer, IPullProducer})
	AssertionError
```

Reply:
Further debug reveals what `providedBy` is:

```
2023-02-17 15:19:50+0000 [-] 2023-02-17 15:19:50,274 - tests.server - 210 - DEBUG - GET-0 - type(producer)=<class 'twisted.protocols.basic.FileSender'>, provides: classImplements(FileSender, IProducer)
```

Reply:
I think this is just a bug in Twisted; AFAICS `FileSender` implements both IPushProducer and IPullProduced, judging by its method signatures.

https://docs.twistedmatrix.com/en/stable/api/twisted.protocols.basic.FileSender.html#resumeProducing

Reply:
That's...wildly annoying. 😢

Reply:
I think I'll need to just do an unchecked `cast` then -- do you see any other options? (I'll file an upstream bug.)

Reply:
Could do a cast, backed up at runtime by and do a `hasattr` check for the methods we use?

It looks like this is https://github.com/matrix-org/synapse/blob/f2d22a77a106724c66086794e49213556e937b27/tests/server.py#L216 and https://github.com/matrix-org/synapse/blob/f2d22a77a106724c66086794e49213556e937b27/tests/server.py#L230

Reply:
I think I'm just going to `cast` and leave a `TODO` for now?

Reply:
SGTM :+1:

Reply:
See twisted/twisted#11811 for an upstream bug.